File: src/java/leetcode/th234/Palindrome.java
package java.leetcode.th234;

/**
 * @fileName: Palindrome.java
 * @author: wyli
 * @createTime: 2023/9/10 21:44
 * @description:
 */
public class Palindrome {
}


class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}


/**
 * https://leetcode.cn/problems/palindrome-linked-list/?envType=study-plan-v2&envId=didi-2023-fall-sprint
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if (head == null){
            return false;
        }
        ListNode cur = head;
        int len = 1;
        while (cur.next != null){
            cur = cur.next;
            len ++;
        }
        int[] arr = new int[len];
        cur = head;
        for (int i = 0; i < len; i++) {
            arr[i] = cur.val;
            cur = cur.next;
        }
        int low = 0;
        int high = len - 1;
        while (low < high){
            if (arr[low] != arr[high]){
                return false;
            }
            low ++;
            high --;
        }
        return true;
    }
}

File: src/cpp/leetcode/th234/Palindrome.cpp
#include<iostream>
#include<vector>
#include<string>

using namespace std;

struct ListNode
{
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode* next) : val(x), next(next) {}
};

class Solution {
public:
    bool isPalindrome(ListNode* head) {
        if (head == nullptr|| head->next == nullptr)
        {
            return true;
        }
        ListNode* pre = head;
        ListNode* cur = head;
        vector<int> arr;
        arr.push_back(pre->val);
        int len = 0;
        while (cur != nullptr)
        {
            len ++;
            arr.push_back(pre->val);
            cur = cur->next;
            pre = pre->next;
            if (cur == nullptr)
            {
                break;
            }
            len ++;
            cur = cur->next;
        }
        int i = arr.size() - 1;
        if (len % 2 != 0)
        {
            i--;
        }
        while (i >= 0 && pre != nullptr)
        {
            if (arr[i--] != pre->val)
            {
                return false;
            }
            pre = pre->next;
        }
        return true;
    }
};

